param/parameterized.py

```diff
@@ -1957,45 +2164,80 @@ def outputs(self_):
                     outputs[name] = (otype, method, idx)
         return outputs
 
+    def _spec_to_obj(self_, spec, dynamic=True):
+        """
+        Resolves dependency specifications into explicit parameter
+        dependencies and dynamic dependencies to be resolved when the
+        sub-object whose parameters are being depended on is defined.
 
-    def _spec_to_obj(self_,spec):
-        # TODO: when we decide on spec, this method should be
-        # rewritten
+        During class creation dynamic=False which means sub-object
+        dependencies are not resolved. At instance creation and
+        whenever a sub-object is set on an object this method will be
+        invoked to determine whether the dependency is available.
 
+        For sub-object dependencies we also return dependencies for
+        every part of the path, e.g. for a dependency specification
+        like "a.b.c" we return dependencies for sub-object "a" and the
+        sub-sub-object "b" in addition to the dependency on the actual
+        parameter "c" on object "b". This is to ensure that if a
+        sub-object is swapped out we are notified and can update the
+        dynamic dependency to the new object. Even if a sub-object
+        dependency can only partially resolved, e.g. if object "a"
+        does not yet have a sub-object "b" we must watch for changes
+        to "b" on sub-object "a".
+        """
```

Comment:
That helps clarify things a lot, thanks! That said, I would expect a single list of dependencies to be returned given this docstring but I see the actual return is two lists, `(deps, deferred)`. What is the difference and what is meant by `deferred`?

Reply:
Thought the first paragraph captured that, so suggestions for improving that welcome:

> Resolves dependency specifications into explicit parameter dependencies and dynamic dependencies to be resolved when the sub-object whose parameters are being depended on is defined.

In other words `deferred` are dependency specs that have not yet been resolved. Either because its class creation time and we don't want to resolve them yet or because they cannot be (fully) resolved because the subobject(s) they are referencing are not yet defined.

Reply:
Can rename deferred to `dynamic_deps` (I have done that in most other places).